apps/showcase/src/components/training/training.component.ts

```diff
@@ -0,0 +1,238 @@
+  dynamicContent: {
+    /** HTML Content - corresponds to the step instructions */
+    htmlContent: WritableSignal<string>;
+    /** Exercise project of the step */
+    project: WritableSignal<TrainingProject | null>;
+    /** Solution of the exercise project of the step */
+    solutionProject: WritableSignal<TrainingProject | null>;
+  };
+};
+
+/** Resources to get file content */
+type Resource = {
+  /** Resource path */
+  path: string;
+  /** Resource content */
+  content: string;
+};
+
+@Component({
+  selector: 'o3r-training',
+  standalone: true,
+  imports: [
+    AsyncPipe,
+    CopyTextPresComponent,
+    DynamicContentModule,
+    FormsModule,
+    JsonPipe,
+    NgbAccordionModule,
+    NgComponentOutlet,
+    TrainingStepPresComponent,
+    NgbDropdown,
+    NgbDropdownToggle,
+    NgbDropdownMenu,
+    NgbDropdownItem
+  ],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  templateUrl: './training.component.html',
+  styleUrl: './training.component.scss'
+})
+export class TrainingComponent implements OnInit {
+  /**
+   * Index of the training step to display
+   */
+  public currentStepIndex = 0;
+  public showInstructions = true;
+  public showSolution = false;
+  /**
+   * Training steps that
+   */
+  public steps: WritableSignal<TrainingStep[]> = signal([]);
+  public stepNames = computed(() =>
+    this.steps().map((step) => step.description.title)
+  );
+  private readonly dynamicContentService = inject(DynamicContentService);
+
+  @Input()
+  public trainingPath = '';
+  @Input()
+  public title = '';
+
+  private async loadStepContent(step: TrainingStep) {
+    if (!step.dynamicContent.htmlContent()) {
+      void this.loadResource(step.description.htmlContentUrl).then((content) =>
+        step.dynamicContent.htmlContent.set(content)
+      );
+    }
+    const fileConfiguration = step.description.filesConfiguration;
+    if (fileConfiguration?.urls && !step.dynamicContent.project()?.files) {
+      await this.updateStepDynamicContent(step, Object.entries(fileConfiguration.urls).map(([path, url])=> ({path, url})));
+    }
+    if (fileConfiguration?.solutionUrls && !step.dynamicContent.solutionProject()?.files) {
+      await this.updateStepDynamicContent(step, [
+        ...Object.entries(fileConfiguration.urls).map(([path, url]) => ({path, url})),
+        ...Object.entries(fileConfiguration.solutionUrls).map(([path, url]) => ({path, url}))
+      ],
+      true);
+    }
+  }
+
+  private async loadSteps() {
+    const program: TrainingStepDescription[] = JSON.parse(await this.loadResource(`${this.trainingPath}/program.json`));
+    const stepsToLoad: TrainingStep[] = [];
+    program.forEach((desc) => {
+      const step = {
+        description: desc,
+        dynamicContent: {
+          htmlContent: signal(''),
+          project: signal(null),
+          solutionProject: signal(null)
+        }
+      };
+      void this.loadStepContent(step);
+      stepsToLoad.push(step);
```

Comment:
Are you sure you will push it in the right order as you are not awaiting the promise ?

Reply:
No issue here, loadStepContent just fill the content, it does not create anything, it will follow the order of the program object

Reply:
It will just update the dynamicContent signals